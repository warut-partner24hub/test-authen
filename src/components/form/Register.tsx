import { useState, useEffect, useCallback } from "react";
import { useForm, SubmitHandler } from "react-hook-form";
import { z } from "zod";
import { zodResolver } from "@hookform/resolvers/zod";
import zxcvbn from "zxcvbn";
import Inputs from "../input/Inputs";
import { toast, ToastContainer } from "react-toastify";
import axios from "axios";
import React from "react";
import { useRouter } from "next/router";

const FormSchema = z
  .object({
    first_name: z
      .string()
      .min(1, "First name is required")
      .min(4, "First name must be more than 4 characters")
      .max(16, "First Name should not be more than 16 characters")
      .regex(/^[a-zA-Z]+$/, "No special character"),
    last_name: z
      .string()
      .min(1, "Last name is required")
      .min(4, "Last name must be more than 4 characters")
      .max(16, "Last Name should not be more than 16 characters")
      .regex(/^[a-zA-Z]+$/, "No special character"),
    email: z.string().email("Please enter a valid email"),
    password: z
      .string()
      .min(1, "Password is required")
      .min(6, "Password must be more than 6 characters")
      .max(16, "Password should not be more than 16 characters"),
    confirm_password: z.string().min(1, "Confirm password is required"),
  })
  .refine((data) => data.password === data.confirm_password, {
    message: "Passwords do not match",
    path: ["confirm_password"],
  });

type FormSchemaType = z.infer<typeof FormSchema>;

const Register: React.FunctionComponent = React.memo(() => {
  const [passwordScore, setPasswordScore] = useState(0);
  const {
    register,
    handleSubmit,
    watch,
    reset,
    formState: { errors, isSubmitting },
  } = useForm<FormSchemaType>({
    resolver: zodResolver(FormSchema),
  });
  const router = useRouter();

  const onSubmit: SubmitHandler<FormSchemaType> = useCallback(
    async (data) => {
      try {
        const response = await axios.post("/api/auth/signup", data);
        reset();
        toast.success(response.data.message);
        setTimeout(() => router.push("/auth"), 3000);
      } catch (error: any) {
        if (error.response?.data?.message) {
          toast.error(error.response.data.message);
        } else {
          toast.error("Registration failed. Please try again.");
        }
      }
    },
    [reset, router]
  );

  const validatePassword = useCallback(() => {
    const password = watch("password");
    return zxcvbn(password || "").score;
  }, [watch]);

  useEffect(() => {
    setPasswordScore(validatePassword());
  }, [validatePassword]);

  return (
    <>
      <ToastContainer />
      <div className="flex items-center justify-center bg-gray-100 min-h-screen">
        <div className="bg-white p-8 rounded shadow-md w-full max-w-md">
          <h1 className="text-2xl font-bold mb-6 text-center">Register Form</h1>
          <form
            className="space-y-6"
            onSubmit={handleSubmit(onSubmit)}
            aria-label="Register form"
          >
            <Inputs
              name="first_name"
              label="First Name"
              type="text"
              placeholder="First Name"
              register={register}
              error={errors.first_name?.message}
              disable={isSubmitting}
            />
            <Inputs
              name="last_name"
              label="Last Name"
              type="text"
              placeholder="Last Name"
              register={register}
              error={errors.last_name?.message}
              disable={isSubmitting}
            />
            <Inputs
              name="email"
              label="Email"
              type="text"
              placeholder="Email"
              register={register}
              error={errors.email?.message}
              disable={isSubmitting}
            />
            <Inputs
              name="password"
              label="Password"
              type="password"
              placeholder="Password"
              register={register}
              error={errors.password?.message}
              disable={isSubmitting}
            />
            {watch("password")?.length > 0 && (
              <div className="flex mt-2" role="progressbar">
                {Array.from(Array(5).keys()).map((span, i) => (
                  <span className="w-1/5 px-1" key={i}>
                    <div
                      className={`h-2 ${
                        passwordScore <= 2
                          ? "bg-red-500"
                          : passwordScore < 4
                          ? "bg-yellow-500"
                          : "bg-green-500"
                      }`}
                    ></div>
                  </span>
                ))}
              </div>
            )}
            <Inputs
              name="confirm_password"
              label="Confirm Password"
              type="password"
              placeholder="Confirm Password"
              register={register}
              error={errors.confirm_password?.message}
              disable={isSubmitting}
            />
            <div className="flex flex-row items-center text-sm">
              <p className="mr-1">Already have an account?</p>
              <button
                type="button"
                className="text-blue-500 cursor-pointer underline focus:outline-none"
                onClick={() => router.push("/auth")}
                aria-label="Sign in"
              >
                Sign in
              </button>
            </div>
            <button
              type="submit"
              className="mt-4 w-full py-2 bg-blue-500 text-white font-semibold rounded hover:bg-blue-600 transition duration-200 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-opacity-50"
              disabled={isSubmitting}
            >
              {isSubmitting ? "Submitting..." : "Submit"}
            </button>
          </form>
        </div>
      </div>
    </>
  );
});

Register.displayName = "Register";

export default Register;
