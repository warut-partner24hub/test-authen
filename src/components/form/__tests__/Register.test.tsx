import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import Register from "../Register";
import { ToastContainer } from "react-toastify";
import axios from "axios";
import { useRouter } from "next/router";

jest.mock("axios");

jest.mock("next/router", () => ({
  useRouter: jest.fn(),
}));

test("Register", () => {
  const { container } = render(<Register />);
  expect(container).toMatchSnapshot();
});

describe("Register Form", () => {
  beforeEach(() => {
    (useRouter as jest.Mock).mockReturnValue({
      push: jest.fn(),
    });
  });

  it("renders Register form correctly", () => {
    render(
      <>
        <ToastContainer />
        <Register />
      </>
    );

    expect(screen.getByLabelText("First Name")).toBeInTheDocument();
    expect(screen.getByLabelText("Last Name")).toBeInTheDocument();
    expect(screen.getByLabelText("Email")).toBeInTheDocument();
    expect(screen.getByLabelText("Password")).toBeInTheDocument();
    expect(screen.getByLabelText("Confirm Password")).toBeInTheDocument();
  });

  it("shows error messages for invalid input", async () => {
    render(
      <>
        <ToastContainer />
        <Register />
      </>
    );

    fireEvent.click(screen.getByText("Submit"));

    await waitFor(() => {
      expect(screen.getByText("First name is required")).toBeInTheDocument();
      expect(screen.getByText("Last name is required")).toBeInTheDocument();
      expect(
        screen.getByText("Please enter a valid email")
      ).toBeInTheDocument();
      expect(screen.getByText("Password is required")).toBeInTheDocument();
      expect(
        screen.getByText("Confirm password is required")
      ).toBeInTheDocument();
    });
  });

  it("submits the form with valid input", async () => {
    const mockedAxios = axios as jest.Mocked<typeof axios>;
    mockedAxios.post.mockResolvedValue({
      data: { message: "Register Success!!" },
    });

    render(
      <>
        <ToastContainer />
        <Register />
      </>
    );

    fireEvent.change(screen.getByLabelText("First Name"), {
      target: { value: "John" },
    });
    fireEvent.change(screen.getByLabelText("Last Name"), {
      target: { value: "Corner" },
    });
    fireEvent.change(screen.getByLabelText("Email"), {
      target: { value: "john@example.com" },
    });
    fireEvent.change(screen.getByLabelText("Password"), {
      target: { value: "password123" },
    });
    fireEvent.change(screen.getByLabelText("Confirm Password"), {
      target: { value: "password123" },
    });

    fireEvent.click(screen.getByText("Submit"));

    await waitFor(() => {
      expect(mockedAxios.post).toHaveBeenCalledTimes(1);
      expect(mockedAxios.post).toHaveBeenCalledWith("/api/auth/signup", {
        first_name: "John",
        last_name: "Corner",
        email: "john@example.com",
        password: "password123",
        confirm_password: "password123",
      });
    });

    await waitFor(() =>
      expect(screen.getByText("Register Success!!")).toBeInTheDocument()
    );
  });

  it("shows error message for existing email", async () => {
    const mockedAxios = axios as jest.Mocked<typeof axios>;
    mockedAxios.post.mockRejectedValue({
      response: { data: { message: "Email already exists" } },
    });

    render(
      <>
        <ToastContainer />
        <Register />
      </>
    );

    fireEvent.change(screen.getByLabelText("First Name"), {
      target: { value: "John" },
    });
    fireEvent.change(screen.getByLabelText("Last Name"), {
      target: { value: "Michel" },
    });
    fireEvent.change(screen.getByLabelText("Email"), {
      target: { value: "john@example.com" },
    });
    fireEvent.change(screen.getByLabelText("Password"), {
      target: { value: "password123" },
    });
    fireEvent.change(screen.getByLabelText("Confirm Password"), {
      target: { value: "password123" },
    });

    fireEvent.click(screen.getByText("Submit"));

    await waitFor(() => {
      expect(mockedAxios.post).toHaveBeenCalledTimes(1);
    });

    await waitFor(() =>
      expect(screen.getByText("Email already exists")).toBeInTheDocument()
    );
  });
});
