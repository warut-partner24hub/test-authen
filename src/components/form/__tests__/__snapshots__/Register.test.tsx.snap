// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="flex items-center justify-center bg-gray-100 min-h-screen"
  >
    <div
      class="bg-white p-8 rounded shadow-md w-full max-w-md"
    >
      <h1
        class="text-2xl font-bold mb-6 text-center"
      >
        Register Form
      </h1>
      <form
        aria-label="Register form"
        class="space-y-6"
      >
        <div
          class="w-full"
        >
          <label
            class="block text-sm font-medium text-gray-700 mb-1"
            for="first_name"
          >
            First Name
          </label>
          <div
            class="relative"
          >
            <input
              class="block w-full px-4 py-2 border rounded border-gray-300 focus:outline-none focus:ring-2 focus:ring-blue-500 bg-white"
              id="first_name"
              name="first_name"
              placeholder="First Name"
              type="text"
            />
          </div>
        </div>
        <div
          class="w-full"
        >
          <label
            class="block text-sm font-medium text-gray-700 mb-1"
            for="last_name"
          >
            Last Name
          </label>
          <div
            class="relative"
          >
            <input
              class="block w-full px-4 py-2 border rounded border-gray-300 focus:outline-none focus:ring-2 focus:ring-blue-500 bg-white"
              id="last_name"
              name="last_name"
              placeholder="Last Name"
              type="text"
            />
          </div>
        </div>
        <div
          class="w-full"
        >
          <label
            class="block text-sm font-medium text-gray-700 mb-1"
            for="email"
          >
            Email
          </label>
          <div
            class="relative"
          >
            <input
              class="block w-full px-4 py-2 border rounded border-gray-300 focus:outline-none focus:ring-2 focus:ring-blue-500 bg-white"
              id="email"
              name="email"
              placeholder="Email"
              type="text"
            />
          </div>
        </div>
        <div
          class="w-full"
        >
          <label
            class="block text-sm font-medium text-gray-700 mb-1"
            for="password"
          >
            Password
          </label>
          <div
            class="relative"
          >
            <input
              class="block w-full px-4 py-2 border rounded border-gray-300 focus:outline-none focus:ring-2 focus:ring-blue-500 bg-white"
              id="password"
              name="password"
              placeholder="Password"
              type="password"
            />
          </div>
        </div>
        <div
          class="w-full"
        >
          <label
            class="block text-sm font-medium text-gray-700 mb-1"
            for="confirm_password"
          >
            Confirm Password
          </label>
          <div
            class="relative"
          >
            <input
              class="block w-full px-4 py-2 border rounded border-gray-300 focus:outline-none focus:ring-2 focus:ring-blue-500 bg-white"
              id="confirm_password"
              name="confirm_password"
              placeholder="Confirm Password"
              type="password"
            />
          </div>
        </div>
        <div
          class="flex flex-row items-center text-sm"
        >
          <p
            class="mr-1"
          >
            Already have an account?
          </p>
          <button
            aria-label="Sign in"
            class="text-blue-500 cursor-pointer underline focus:outline-none"
            type="button"
          >
            Sign in
          </button>
        </div>
        <button
          class="mt-4 w-full py-2 bg-blue-500 text-white font-semibold rounded hover:bg-blue-600 transition duration-200 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-opacity-50"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>
</div>
`;
